participant/helpers.py

```diff
@@ -0,0 +1,159 @@
+                self.get_subquery_value(
+                    participant_actions_qs, "email_opened_attachment"
+                ),
+                0,
+            ),
+            participant_count=self.get_subquery_value(
+                exercise_participants_qs, "participant_count"
+            ),
+            training_link_clicked=self.get_subquery_value(
+                participant_actions_qs, "clicked_training_link"
+            ),
+        ).values(
+            "id",
+            "title",
+            "trial_version",
+            "pos_reported",
+            "pos_deleted",
+            "neg_clicked_link",
+            "neg_entered_detail",
+            "neg_replied_to_phishing_email",
+            "neg_opened_attachment",
+            "participant_count",
+            "training_link_clicked",
+        )
+
+        csv_rows = list(rows_qs)
+
+        for row in csv_rows:
+            row["emails_opened"] = email_type_count_map[row["id"]]["regular_emails"]
+            row["phishing_emails_opened"] = email_type_count_map[row["id"]][
+                "phishing_emails"
+            ]
+
+        return csv_columns, csv_rows
+
+    @staticmethod
+    def get_subquery_value(queryset, value):
+        return Subquery(queryset.values(value)[:1], output_field=CharField())
+
+    @staticmethod
+    def get_action_count(action):
+        return Count(
+            "id", filter=Q(actionlog__name="action_type", actionlog__value=action)
+        )
+
+    @staticmethod
+    def get_emails_count_map(exercise_qs):
+        emails_count_map = {}
+        for exercise in exercise_qs:
+            # Participants actions based on the exercise
+            participant_actions_queryset = ParticipantAction.objects.filter(
+                participant__exercise=exercise
+            )
+            # serializing actions to filter only the ones corresponding to emails opened
+            serialized_actions = ParticipantActionSerializer(
+                participant_actions_queryset, many=True
+            ).data
+
+            # Filtering email opened ids
+            email_uuids = {
+                action["action_details"]["email_id"]
+                for action in serialized_actions
+                if (action["action_details"].get("email_id"))
+                and (action["action_details"].get("action_type") == "email_opened")
+            }
+
+            # Getting opened emails to verify if they are phishing or regular ones
+            exercise_scoped_emails = exercise_models.ExerciseEmail.objects.filter(
+                id__in=email_uuids
+            )
+
+            phishing_emails_count = len(
+                exercise_scoped_emails.filter(
+                    phish_type=exercise_models.EXERCISE_EMAIL_PHISH
+                )
+            )
+            regular_emails_count = len(
+                exercise_scoped_emails.filter(
+                    phish_type=exercise_models.EXERCISE_EMAIL_REGULAR
+                )
+            )
```

Comment:
just do `.count()` on these pls instead of `len()`

Reply:
Done